vmse2000.py

```diff
@@ -434,6 +475,51 @@ def run(self):
             self.stop_threads()
             self.clean_up()
 
+    def filepop(self, filename):
+        """Pop the last non-empty line from a file and return it. If the file is empty, return None."""
+        if not filename:
+            return None
+        with open(filename, "r+") as f:
+            f.seek(0, os.SEEK_END)
+            size = f.tell()
+            if size == 0:
+                return None
+            # skip trailing newlines
+            pos = size - 1
+            f.seek(pos, os.SEEK_SET)
+            c = f.read(1)
+            while c == "\n" or c == "\r":
+                if pos == 0:
+                    return None
+                pos -= 1
+                f.seek(0, os.SEEK_SET)
+                c = f.read(1)
+            if pos == 0:
+                # no more (none-empty) lines
+                return None
+            line_end = pos + 1
+            # find the start of the last line
+            while pos > 0:
+                pos -= 1
+                f.seek(pos, os.SEEK_SET)
+                c = f.read(1)
+                if c == "\n" or c == "\r":
+                    pos += 1
+                    break
+            line_start = pos
+            # print("line_start:", line_start, "line_end:", line_end)
+            f.seek(line_start, os.SEEK_SET)
+            line = f.read(line_end - line_start)
+            # print("popped line:", line)
+            # truncate the file
+            f.truncate(line_start)
+            return line.strip()
```

Comment:
While I applaud the effort of doing all the line wrangling I think that this is quite sensitive to interruptions, i.e. the chance of everything breaking when *some* error (reboot, power loss, etc.) occurs is quite high.

Maybe it'd be best to have a second index file that simply stores a number (append). This way we can easily recover and it's only one append-write. Loading would be one seek + one line read (i.e. the equivalent of `tail -1 foo` for loading and `echo $i >> foo` for storing).

Reply:
I do have a try/catch around that in the next PR